### PR TITLE
File list

### PR DESCRIPTION
**Pull request by mathjeff**
Support for a custom list of Blueprints files to parse

### CODE REVIEW

File: bootstrap/command.go [hunk 1]
```diff
@@ -89,8 +110,13 @@ func Main(ctx *blueprint.Context, config interface{}, extraNinjaFileDeps ...stri
...
 
-	if flag.NArg() != 1 {
-		fatalf("no Blueprints file specified")
+	filesToParse, err := listFiles()
+	if err != nil {
+		if flag.NArg() != 1 {
+			fatalf("no Blueprints file specified, and "+
+				"could not read list file because: %v\n", err.Error())
+		}
+		filesToParse = []string{flag.Arg(0)}
```
Comment (mathjeff) [inline bootstrap/command.go]: I suspect that we don't want to be passing both a list of files via listFiles() and also a root file via flag.Arg(0) . I'm still investigating a better way make that work

**Comment by colincross** [comment]
These patches are on top of the Android history, can you rebase them on top of upstream master?

File: context.go [hunk 2]
```diff
@@ -689,9 +690,9 @@ loop:
 	return deps, errs
 }
 
-type FileHandler func(*parser.File)
+type FileVisitor func(*parser.File)
```
Comment (colincross) [inline context.go]: Handler makes more sense than Visitor to me, Visitor implies walking something that is already parsed to me.  The implementation above is also still called handleOneFile.

File: context.go [hunk 3]
```diff
@@ -514,17 +514,18 @@ func (c *Context) SetAllowMissingDependencies(allowMissingDependencies bool) {
 	c.allowMissingDependencies = allowMissingDependencies
 }
 
-// Parse parses a single Blueprints file from r, creating Module objects for
-// each of the module definitions encountered.  If the Blueprints file contains
-// an assignment to the "subdirs" variable, then the subdirectories listed are
-// searched for Blueprints files returned in the subBlueprints return value.
-// If the Blueprints file contains an assignment to the "build" variable, then
-// the file listed are returned in the subBlueprints return value.
+// parseOne parses a single Blueprints file from <reader>, creating Module
```
Comment (colincross) [inline context.go]: godoc style doesn't use <> around parameters.

File: context.go [hunk 4]
```diff
@@ -783,11 +785,15 @@ func (c *Context) MockFileSystem(files map[string][]byte) {
...
-func (c *Context) parseBlueprintsFile(filename string, scope *parser.Scope, rootDir string,
+// parseOneCh parses a single Blueprints file, and sends results through the provided channels
+//
+// Errors are returned through errsCh.
+// Any defined modules are returned through modulesCh.
+// Any sub-Blueprints files are returned through blueprintsCh.
+// Any dependencies on Blueprints files or directories are returned through depsCh.
+// When processing completes (and all other channel responses have been sent successfully),
+// an empty struct is passed into doneCh.
+func (c *Context) parseOneCh(filename string, scope *parser.Scope, rootDir string,
```
Comment (colincross) [inline context.go]: It's not parsing a channel, its parsing a blueprints file

File: bootstrap/command.go [hunk 5]
```diff
@@ -58,6 +61,27 @@ func init() {
 	flag.BoolVar(&runGoTests, "t", false, "build and run go tests during bootstrap")
 }
 
+func listFiles(srcDir string) (filePaths []string, err error) {
+	listFile := ModulesFile
```
Comment (colincross) [inline bootstrap/command.go]: Take listFile as a parameter so that only Main is referencing the global variables

File: bootstrap/command.go (hunk 1 above)
Comment (colincross) [reply bootstrap/command.go]: Yeah, we might delete the root Android.bp file.  On the other hand, people will expect it to exist so maybe we will replace it with one containing a comment that explains the new system, in which case we can punt on supporting invoking blueprint without a root file until later.

File: context.go [hunk 6]
```diff
@@ -641,15 +652,33 @@ func (c *Context) WalkBlueprintsFiles(rootFile string, visitor FileVisitor) (dep
 			c.parseOneCh(blueprint.string, blueprint.Scope, rootDir,
 				errsCh, fileCh, blueprintsCh, depsCh)
 
+			descendants, hasDescendants := descendantsMap[blueprint.string]
+			if hasDescendants {
+				for _, descendant := range descendants {
+					blueprintsCh <- stringAndScope{descendant, parser.NewScope(blueprint.Scope)}
```
Comment (colincross) [inline context.go]: I think it makes more sense for the outer goroutine to handle the descendants.  I suggest removing blueprintCh, and returning the blueprint struct on doneCh.

File: context.go [hunk 7]
```diff
@@ -1399,6 +1423,57 @@ func (c *Context) addInterVariantDependency(origModule *moduleInfo, tag Dependen
 	atomic.AddUint32(&c.depsModified, 1)
 }
 
+// parentFile returns the path of a file with the same name but in a parent directory
+// For example, if path = "a/b/Blueprints", then parent = "a/Blueprints"
+func parentFile(path string) (parent string, hasParent bool) {
```
Comment (colincross) [inline context.go]: Would it be easier to do the ancestor mapping on directories and ignore the file name?

File: context.go [hunk 8]
```diff
@@ -1399,6 +1423,57 @@ func (c *Context) addInterVariantDependency(origModule *moduleInfo, tag Dependen
...
+		return "", false
+	}
+	return parentFile, true
+}
+
+// findBlueprintAncestors returns two maps linking parents to children
+// For example, if paths = []string{"a/b/c/Blueprints", "a/Blueprints"},
+// then ancestors = {"a/b/c/Blueprints":"a/Blueprints"}
+// and descendants = {"a/Blueprints", ["a/b/c/Blueprints"]}
+func findBlueprintAncestors(paths []string) (ancestors map[string]string, descendants map[string][]string) {
```
Comment (colincross) [inline context.go]: Returning descendants and a list of paths that have no ancestors would be more useful.  Or just return descendants, and put paths with no ancestors as descendants of "".

File: context.go (hunk 6 above)
Comment (mathjeff) [reply context.go]: Does this mean to drop support for "subdirs="?

Comment (colincross) [reply context.go]: I guess leave blueprintCh, but also return blueprint on doneCh so the outer one can handle descendants?  Or leave it this way if you think its cleaner.

File: context.go (hunk 8 above)
Comment (mathjeff) [reply context.go]: thanks

File: context.go (hunk 7 above)
Comment (mathjeff) [reply context.go]: done

File: context.go (hunk 6 above)
Comment (mathjeff) [reply context.go]: done

File: bootstrap/command.go (hunk 5 above)
Comment (mathjeff) [reply bootstrap/command.go]: thanks

File: context.go (hunk 4 above)
Comment (mathjeff) [reply context.go]: ok

File: context.go (hunk 2 above)
Comment (mathjeff) [reply context.go]: done

File: context.go [hunk 9]
```diff
@@ -642,15 +733,31 @@ func (c *Context) WalkBlueprintsFiles(rootFile string, visitor FileVisitor) (dep
...
+		if hasDescendants {
+			for _, descendant := range descendants {
+				foundParseableBlueprint(stringAndScope{descendant, parser.NewScope(blueprint.Scope)})
+			}
+		}
+	}
 
-	var pending []stringAndScope
+	// begin parsing any files that have no ancestors
+	for _, path := range descendantsMap[""] {
```
Comment (colincross) [inline context.go]: I'm not sure if its better, but this could be startParseDescendants(stringAndScope{"", parser.NewScope(nil)})

File: context.go [hunk 10]
```diff
@@ -669,18 +776,18 @@ loop:
...
-				continue
-			}
-			startParseBlueprintsFile(blueprint)
-		case <-doneCh:
+			foundParseableBlueprint(blueprint)
+		case blueprint := <-doneCh:
 			activeCount--
-			if len(pending) > 0 {
-				startParseBlueprintsFile(pending[len(pending)-1])
+			deps = append(deps, blueprint.string)
```
Comment (colincross) [inline context.go]: Can this go in startParseDescendants instead?

**Comment by mathjeff** [comment]
Hmm it looks like the latest version of this patchset results in slower execution than without it, so we'll want to wait to actually merge it until that's fixed

File: context.go [hunk 11]
```diff
@@ -792,51 +906,21 @@ func (c *Context) parseOne(rootDir, filename string, reader io.Reader,
 	}
 	file.Name = relBlueprintsFile
 
-	subdirs, subdirsPos, err := getLocalStringListFromScope(scope, "subdirs")
```
Comment (colincross) [inline context.go]: Are these deletions the only thing required to remove subdirs support?  If so, lets leave them in and then we can figure out what we want to do with legacy subdirs support in a separate patch.  Using both file list and subdirs together shouldn't cause a problem because startParseBlueprintsFile checks if the file has already been parsed.

File: context.go [hunk 12]
```diff
@@ -1417,6 +1465,43 @@ func (c *Context) addInterVariantDependency(origModule *moduleInfo, tag Dependen
...
+		if alreadyFound {
+			return nil, fmt.Errorf("Found two Blueprint files in directory %v : %v and %v", dir, filesByDir[dir], path)
+		}
+		filesByDir[dir] = path
+	}
+	// generate the descendants map
+	descendants = make(map[string][]string, len(filesByDir))
+	for childDir, childFile := range filesByDir {
+		prevAncestorDir := childDir
+		for {
```
Comment (colincross) [inline context.go]: Not sure if its better, but this could be:
for ancestorDir := filepath.Dir(prevAncestorDir); ancestorDir != prevAncestorDir {

File: context.go [hunk 13]
```diff
@@ -1417,6 +1465,43 @@ func (c *Context) addInterVariantDependency(origModule *moduleInfo, tag Dependen
...
+			ancestorDir := filepath.Clean(filepath.Dir(prevAncestorDir))
```
Comment (colincross) [inline context.go]: Dir will always return a clean path, no need to Clean it again.

File: context.go (hunk 9 above)
Comment (mathjeff) [reply context.go]: ok

File: context.go (hunk 10 above)
Comment (mathjeff) [reply context.go]: done

File: context.go (hunk 11 above)
Comment (mathjeff) [reply context.go]: done

File: context.go (hunk 13 above)
Comment (mathjeff) [reply context.go]: thanks

File: context.go (hunk 12 above)
Comment (mathjeff) [reply context.go]: It does also need to add childFile as a descendant of descendants[""] if it reaches the root directory. Could you clarify on the syntax that you would like, or shall we just leave it?

File: bootstrap/command.go [hunk 14]
```diff
@@ -148,7 +148,7 @@ func Main(ctx *blueprint.Context, config interface{}, extraNinjaFileDeps ...stri
 	deps = append(deps, extraDeps...)
```
Comment (mathjeff) [inline bootstrap/command.go]: I suspect a dependency somewhere, possibly this <deps> , needs to include the $BLUEPRINT_LIST_FILE in order to rerun the right invocations of Blueprint when a new Android.bp is created.

File: context.go (hunk 11 above)
Comment (mathjeff) [reply context.go]: Hmm it turns out that there exists a symlink (not in a .git repo, but specified by a manifest) named Android.bp pointing to an Android.bp in a subdirectory, which has a nonempty value for "subdirs".

Previously the parent Android.bp was the one that was getting parsed by Blueprint, so the relative file path listed in the child is supposed to be interpreted relative to the parent. Then when this change causes Blueprint to search for Android.bp files, it finds the child Android.bp and looks for the target of the subdirs relative to the child, and fails to open the result.

Do you think that that makes it worthwhile to remove subdirs support at the same time, or whether we want to fix the Android.bp files (probably involving renaming them and updating some manifests to repoint the symlink) ?

File: bootstrap/command.go (hunk 14 above)
Comment (mathjeff) [reply bootstrap/command.go]: Yeah that ended up being added in the next patchset

**Comment by mathjeff** [comment]
It looks like the Travis build failed because tests/test_tree_tests.sh uses a symlink: blueprint -> ../.. 

It seems to me that the most convenient way to fix it starts with merging https://github.com/google/blueprint/pull/158

**Comment by colincross** [comment]
Merged it an reran travis but it still fails

**Comment by mathjeff** [comment]
Ping!

Does this patch need more review or is it only blocked on merging the removals of various non-working projects?

File: bootstrap/command.go [hunk 15]
```diff
@@ -89,16 +91,19 @@ func Main(ctx *blueprint.Context, config interface{}, extraNinjaFileDeps ...stri
 		defer trace.Stop()
 	}
 
-	if flag.NArg() != 1 {
-		fatalf("no Blueprints file specified")
```
Comment (danw) [inline bootstrap/command.go]: Why are you removing the NArg check?

File: bootstrap.bash [hunk 16]
```diff
@@ -101,11 +106,14 @@ echo "extraArgs = $EXTRA_ARGS" >> $BUILDDIR/.minibootstrap/build.ninja
 echo "builddir = $NINJA_BUILDDIR" >> $BUILDDIR/.minibootstrap/build.ninja
 echo "include $BLUEPRINTDIR/bootstrap/build.ninja" >> $BUILDDIR/.minibootstrap/build.ninja
 
-echo "BLUEPRINT_BOOTSTRAP_VERSION=1" > $BUILDDIR/.blueprint.bootstrap
+echo "BLUEPRINT_BOOTSTRAP_VERSION=2" > $BUILDDIR/.blueprint.bootstrap
 echo "SRCDIR=\"${SRCDIR}\"" >> $BUILDDIR/.blueprint.bootstrap
 echo "BLUEPRINTDIR=\"${BLUEPRINTDIR}\"" >> $BUILDDIR/.blueprint.bootstrap
 echo "NINJA_BUILDDIR=\"${NINJA_BUILDDIR}\"" >> $BUILDDIR/.blueprint.bootstrap
 echo "GOROOT=\"${GOROOT}\"" >> $BUILDDIR/.blueprint.bootstrap
+echo "TOPNAME=\"${TOPNAME}\"" >> $BUILDDIR/.blueprint.bootstrap
```
Comment (danw) [inline bootstrap.bash]: Why does TOPNAME need to be preserved?

File: bootstrap/command.go [hunk 17]
```diff
@@ -161,9 +168,19 @@ func Main(ctx *blueprint.Context, config interface{}, extraNinjaFileDeps ...stri
...
-		if err != nil {
-			fatalf("error writing depfile: %s", err)
+		if moduleListFile != "" {
+			err := deptools.WriteDepFile(depFile, outFile, deps)
+			if err != nil {
+				fatalf("error writing depfile: %s", err)
+			}
+		} else {
+			// If no file list is provided, then essentially every directory is a dependency
+			// Rather than emitting a list of every directory, we just skip making the deps file.
```
Comment (danw) [inline bootstrap/command.go]: The depfile contains more than just the Blueprints/Android.bp files, I don't think we can remove it entirely. Or is this just triggering it to just always build?

File: context.go [hunk 18]
```diff
@@ -106,7 +107,9 @@ type Context struct {
 	globs    map[string]GlobPath
 	globLock sync.Mutex
 
-	fs pathtools.FileSystem
+	fs                 pathtools.FileSystem
+	sourceFinder       func(baseDir string) ([]string, error)
+	ExtraNinjaFileDeps []string
```
Comment (danw) [inline context.go]: Why not just returns a list of deps from SetModuleListFile like all of the other ctx functions?

File: context.go [hunk 19]
```diff
@@ -285,6 +289,36 @@ func NewContext() *Context {
 	return ctx
 }
 
+// searches for Blueprints files under baseDir
+func findBlueprintsFromDir(baseDir string) (filePaths []string, err error) {
```
Comment (danw) [inline context.go]: If we're going to implement this here, can we also export a list of dependencies so that the dep file actually works?

File: tests/test_tree_tests.sh [hunk 20]
```diff
@@ -25,39 +26,17 @@ cp -r "${TOP}" "${SRCDIR}/blueprint"
...
 OLDTIME_BOOTSTRAP=$(mtime .bootstrap/build.ninja)
 OLDTIME=$(mtime build.ninja)
 
-sleep 2
-./blueprint.bash
-
-if [ ${OLDTIME} != $(mtime build.ninja) ]; then
-    echo "unnecessary build.ninja regeneration for null build" >&2
-    exit 1
-fi
```
Comment (danw) [inline tests/test_tree_tests.sh]: Why are these tests being removed?

File: bootstrap/command.go (hunk 15 above)
Comment (mathjeff) [reply bootstrap/command.go]: Hmm, that must've been a mistake

File: bootstrap/command.go (hunk 17 above)
Comment (mathjeff) [reply bootstrap/command.go]: This was triggering it to always rebuild, although the nextpatchset takes your suggestion of always generating the listfile, so this code gets removed

File: context.go (hunk 18 above)
Comment (mathjeff) [reply context.go]: thanks

File: context.go (hunk 19 above)
Comment (mathjeff) [reply context.go]: this is removed in the next patchset at your suggestion

File: bootstrap.bash (hunk 16 above)
Comment (mathjeff) [reply bootstrap.bash]: Oops, it wasn't needed for this patchset. It is needed for the next patchset after re-adding the bash search though.

File: tests/test_tree_tests.sh (hunk 20 above)
Comment (mathjeff) [reply tests/test_tree_tests.sh]: they didn't work in this patchset, but the nextpatchset fixes them and keeps them in
